[RFC] main.c improvements

Pyrohh opened the pull request:
Among other things, this addresses usage of `char_u` for character data (#1818). I'm not completely sure if I'm dealing with `-Wconversion` optimally here, so a check on the first commit would be much appreciated.

justinmk commented on src/nvim/main.c:
```diff
@@ -834,20 +808,25 @@ void getout(int exitval)
-    int *idx,           /* index in argument, is incremented */
-    int def                    /* default value */
-)
+/// Gets the integer value of a numeric command line argument if given,
```
Nice doc improvement.

justinmk commented on src/nvim/main.c:
```diff
@@ -834,20 +808,25 @@ void getout(int exitval)
+/// such as '-o10'.
+///
+/// @param p pointer to argument
+/// @param idx pointer to index in argument
```
Should keep note about increment

elmart replied:
Yes. Also, we have the convention to mark such params that can be modified with `@param[out]`.

Pyrohh replied:
Noted

Pyrohh replied:
I'm a little confused after reading this: http://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdparam

Given that `*idx` is a parameter that is modifed, would the following be satisfactory?
```
/// @param[in] p pointer to argument
/// @param[in,out] idx pointer to index in argument
/// @param[in] def default value
```

elmart replied:
Yes. `[in, out]` would be more correct. Note the space after the comma.

Pyrohh replied:
That looks a bit awkward, since the `o` in `out]` is lined up with the `p` above it, but I'll do that for now.

elmart replied:
Well, I in fact would line up all descriptions in a column, after the brackets.

justinmk commented on src/nvim/main.c:
```diff
@@ -2044,13 +2017,11 @@ process_env (
- */
-static int file_owned(char *fname)
+#ifdef UNIX
+/// Checks if we own the file "fname".
```
If user owns file

justinmk commented on src/nvim/main.c:
```diff
@@ -2062,34 +2033,28 @@ static int file_owned(char *fname)
-    char_u *str       /* extra argument or NULL */
-)
+/// Prints the following then exits:
+/// - Nvim's version
```
Looks like those lines were removed

elmart replied:
Yes. It doesn't print version anymore.

Pyrohh replied:
I forgot to update the comment (I'll do so), but I did that on purpose. From the commit message:
```
Don't print version when called. mainerr() only handles errors
related to command line arguments, so the version isn't much help at
all.
```
and the mentioned errors:
```
Unknown option argument
Too many edit arguments
Argument missing after
Garbage after option argument
Too many "+command", "-c command" or "--cmd command" arguments
```

Pyrohh replied:
I just realized you probably already saw the commit message, my mistake.

justinmk commented:
Seems nice. Probably should squash

elmart commented on src/nvim/globals.h:
```diff
@@ -140,7 +140,7 @@ EXTERN int mod_mask INIT(= 0x0);                /* current key modifiers */
  * update_screen().
  */
-EXTERN int cmdline_row;
+EXTERN long cmdline_row;
```
Please note that you can't just change the type. 
You have to review every place where that variable is used to check the new type doesn't introduce problems.
For `cmdline_row`, there are 81 such usages.
Note that the fact that it compiles without errors doesn't mean anything, as many of those usages are in files where -Wconversion is not yet enabled. And, even if -Wconversion was enabled for all files, you still should manually review every usage, because of possible previous explicit casts becoming incorrect (those would not be caught by -Wconversion).
See, for example, ex_getln.c:2462: `windgoto(cmdline_row, 0);`, where `windgoto()` is defined as having an `int` first parameter.

So, going through that messy process would be the only way to do it correctly, if we really wanted to.
But we don't need that. Because, in fact, an int is enough for that variable (you won't have a screen with more than 2Glines). That is also true for `Rows` and `p_ch` (an `int` would be enough for them). But those *must* be `long` (because those are user options, and the way options are implemented by now, all numeric options have to be the same type, which turns out to be `long`). I'm in fact studying possible refactorings of this, to eliminate `long` from the codebase.

Then, how to remove the warning to let us enable -Wconversion?
In short, a safe cast.
This is, check that the substraction is within `int`bounds, and cast to int:
```
  assert(Rows - p_ch >= INT_MIN && Rows - p_ch <= INT_MAX);
  cmdline_row = (int)(Rows - p_ch);
```
In the general case, that wouldn't be completely correct yet, as the substraction could overflow/underflow. So, we'd have to check that substraction doesn't overflow as long, and result is within int bounds.
But we don't need that in this case, as we have extra knowledge (both Rows and p_ch must be positive, and Rows >= p_ch). So, introducing all that extra knowledge in the assert (which also helps us validate our assumptions about the code), it would remain:
```
  assert(p_ch >= 0 && Rows >= p_ch &&  Rows - p_ch <= INT_MAX);
  cmdline_row = (int)(Rows - p_ch);
```
It could still be improved, because it seems that Rows must be between 2 and 1000, and p_ch between 1 and Rows (exclusive). But that could be overkill, as limitation about Rows could be removed. Once both substraction and cast are safe, it's enough for me.

As you see, type refactoring is a tricky thing. Please be extra-picky when dealing with it. ;-)
Hope that helps.

Pyrohh replied:
Thank you, that was very helpful. Is the cast

Pyrohh replied:
Thank you, that was very helpful. Do you know if asserts can be optimized away in such cases?

elmart replied:
In what cases?
Asserts are there in the DEBUG build, but are not included in the RELEASE build.

Pyrohh replied:
> In what cases?

My wording was poor, I meant `in this case`. I was (possibly incorrectly) conflating `assert` with something like `memset_s`. I'm not particularly experienced with C, so these may seem like silly questions.

elmart commented on src/nvim/globals.h:
```diff
@@ -190,7 +190,7 @@ EXTERN int compl_cont_status INIT(= 0);
 EXTERN int cmdmsg_rl INIT(= FALSE);         /* cmdline is drawn right to left */
 EXTERN int msg_col;
-EXTERN int msg_row;
+EXTERN long msg_row;
```
Once the above is done, this can remain int too.

elmart commented:
Nice changes. I particularly like `char_u`-related ones. There is a huge number of casts in the codebase to be removed that way.

Pyrohh commented:
I made the requested changes and squashed. I'm not sure if the warning on line 280 (below the assert) mean anything of importance.

elmart commented on src/nvim/main.c:
```diff
@@ -293,7 +276,8 @@ int main(int argc, char **argv)
 
-  cmdline_row = Rows - p_ch;
+  assert(p_ch >= 0 && Rows >= p_ch && Rows - p_ch <= INT_MAX);
+  cmdline_row = (int)Rows - p_ch;
```
The assert is wrong. I wrote `(int)(Rows - p_ch)`, not `(int)Rows - p_ch`.
What you have written behaves this way:
- Rows is converted to int.
- Before doing substraction, as you have an int and a long (p_ch), the int gets promoted to long again.
- The substraction is done.
- Result of the substraction (long) is going to be assigned to an int var (cmdline_row), loosing precision. Thus the warning.

Pyrohh replied:
I understand, thanks.

elmart commented:
Also note that in the travis build, warnings are treated as errors. You have to be warning-free for the build to pass.

justinmk commented on src/nvim/main.c:
```diff
@@ -119,25 +119,21 @@ typedef struct {
+#define ME_TOO_MANY_ARGS   1
+#define ME_ARG_MISSING     2
+#define ME_GARBAGE         3
+#define ME_EXTRA_CMD       4
```
I think it was useful when these defines lived next to their respective messages.

Pyrohh replied:
How about this?
```diff
--- a/src/nvim/main.c
+++ b/src/nvim/main.c
@@ -120,19 +120,12 @@ typedef struct {
 #endif
 
 // Error messages
-static const char *main_errors[] = {
-  N_("Unknown option argument"),
-  N_("Too many edit arguments"),
-  N_("Argument missing after"),
-  N_("Garbage after option argument"),
-  N_("Too many \"+command\", \"-c command\" or \"--cmd command\" arguments")
-};
+#define ME_UNKNOWN_OPTION  "Unknown option argument"
+#define ME_TOO_MANY_ARGS   "Too many edit arguments"
+#define ME_ARG_MISSING     "Argument missing after"
+#define ME_GARBAGE         "Garbage after option argument"
+#define ME_EXTRA_CMD       "Too many \"+command\", \"-c command\" or \"--cmd command\" arguments"
 
-#define ME_UNKNOWN_OPTION  0
-#define ME_TOO_MANY_ARGS   1
-#define ME_ARG_MISSING     2
-#define ME_GARBAGE         3
-#define ME_EXTRA_CMD       4
 
 /// Performs early initialization.
 ///
@@ -2036,16 +2029,16 @@ static bool file_owned(const char *fname)
 #endif
 
 /// Prints the following then exits:
-/// - An error message main_errors[n]
+/// - An error string err
 /// - A string str if not null
 ///
-/// @param n    error number represented by an ME_* macro
+/// @param err  error string represented by an ME_* macro
 /// @param str  string to append to the primary error message, or NULL
-static void mainerr(int n, const char *str)
+static void mainerr(char *err, const char *str)
 {
   signal_stop();              // kill us with CTRL-C here, if you like
 
-  mch_errmsg(_(main_errors[n]));
+  mch_errmsg(_(err));
   if (str != NULL) {
     mch_errmsg(": \"");
     mch_errmsg(str);
```

justinmk replied:
I think the `N_` is needed for localization. Better to leave it as it was, there is not a compelling reason to change it.

Pyrohh replied:
The [gettext manual](https://www.gnu.org/software/gettext/manual/html_node/xgettext-Invocation.html) states that preprocessor macros are processed, but I can drop this change if you still want to.

justinmk replied:
Your patch above has this:

    #define ME_UNKNOWN_OPTION  "Unknown option argument"

Are you saying gettext will process `"Unknown option argument"` even though it is not wrapped in `N_()`? I did not see anywhere in the manual that says that.

Pyrohh replied:
This should pop up if you `ctrl-f` for it:
```
This implementation of xgettext is able to process a few awkward cases, like strings in preprocessor macros
```
I'm currently trying to verify that's actually what happens.

justinmk replied:
I read that line, but it sounds vague. It also implies that some gettext implementations may not have this feature. 

Anyways, I'm interested from an academic perspective, but even if it does somehow magically localize every `#define`d string, we must be careful not to change things just for change's sake.

Pyrohh replied:
> we must be careful not to change things just for change's sake.
I think there is a clear (albeit small) improvement here, but I understand. It's not worth it for something so small, so I'll revert it.

Pyrohh commented:
> Also note that in the travis build, warnings are treated as errors. You have to be warning-free for the build to pass.

I see, repushed.

Commit:
main.c: Enable -Wconversion

Commit:
main.c: Misc. improvements

- Constify parameters and reduce char_u usage
- Modernize function signatures
- Modernize function comment blocks, rewriting if needed
- Factor out mainerr_arg_missing(); all the other ME_* macros lack
  such a wrapper function.

main_errors[]
  - Remove unneeded parentheses
  - Remove an unused error string and its respective macro

mainerr()
  - Don't print version when called. mainerr() only handles errors
    related to command line arguments, so the version isn't much help at
    all.
  - Changed 'vim' instance to 'nvim'

Misc.
  - Remove duplicate include
  - Replace unneeded 'inttypes.h' include with 'stdint.h'
  - Remove stray comments
  - Remove excessive newlines

Pyrohh commented:
@justinmk Re-inlined the `ME_*

Pyrohh commented:
@justinmk Re-inlined the `ME_*` macros and repushed.

justinmk commented:
Thanks!
